File: weather_backend/routes/weather.py
"""
weather.py
Endpoints /weather et /cities
"""
from fastapi import APIRouter, Query, HTTPException
from services import weather_service
from schemas.weather_schema import CityOut, WeatherTodayOut, WeatherHistoryOut
from utils.helpers import ensure_city_exists, handle_db_error

router = APIRouter()


@router.get("/cities", response_model=list[CityOut])
def list_cities():
    try:
        return weather_service.get_all_cities()
    except ConnectionError as e:
        handle_db_error(e)


@router.get("/weather/today", response_model=list[WeatherTodayOut])
def weather_today(city: str = Query(None, description="Filtrer par ville")):
    try:
        ensure_city_exists(city, weather_service.is_valid_city)
        data = weather_service.get_weather_today(city)
        if not data:
            raise HTTPException(status_code=404, detail="Aucune donnée météo pour aujourd'hui")
        return data
    except ConnectionError as e:
        handle_db_error(e)


@router.get("/weather/history", response_model=list[WeatherHistoryOut])
def weather_history(
    city: str = Query(None),
    date_from: str = Query(None, description="Format YYYY-MM-DD"),
    date_to: str = Query(None, description="Format YYYY-MM-DD"),
):
    try:
        ensure_city_exists(city, weather_service.is_valid_city)
        return weather_service.get_weather_history(city, date_from, date_to)
    except ConnectionError as e:
        handle_db_error(e)

File: weather_frontend/src/api/weather.js
import axios from 'axios'

const API = axios.create({
  baseURL: '/api',
  timeout: 10000
})

// Cities
export const getCities = async () => {
  const { data } = await API.get('/cities')
  return data
}

// Weather Today
export const getWeatherToday = async (city = null) => {
  const params = city ? { city } : {}
  const { data } = await API.get('/weather/today', { params })
  return data
}

// Weather History
export const getWeatherHistory = async (city = null, dateFrom = null, dateTo = null) => {
  const params = {}
  if (city) params.city = city
  if (dateFrom) params.date_from = dateFrom
  if (dateTo) params.date_to = dateTo

  const { data } = await API.get('/weather/history', { params })
  return data
}

// Alerts
export const getAlerts = async (city = null) => {
  const params = city ? { city } : {}
  const { data } = await API.get('/weather/alerts', { params })
  return data
}

export const getAlertsSummary = async (city = null) => {
  const params = city ? { city } : {}
  const { data } = await API.get('/weather/alerts/summary', { params })
  return data
}